extensions/cdc-debezium/src/main/java/com/hazelcast/jet/cdc/impl/CdcSourceP.java

```diff
@@ -48,81 +54,159 @@
         }
+
+        this.clearStateOnReconnect = getClearStateOnReconnect(properties);
+        log(clearStateOnReconnect, "clear state on reconnect", name, logger);
     }
 
-    public void fillBuffer(SourceBuilder.TimestampedSourceBuffer<T> buf) {
+    @Override
+    public boolean complete() {
         if (reconnectTracker.needsToWait()) {
-            return;
+            return false;
         }
-
         if (!isConnectionUp()) {
-            return;
+            return false;
+        }
+        if (snapshotInProgress) {
+            return false;
+        }
+        if (!emitFromTraverser(traverser)) {
+            return false;
         }
 
         try {
+            if (commitPeriod > 0) {
+                long currentTime = System.nanoTime();
+                if (currentTime - lastCommitTime > commitPeriod) {
+                    task.commit();
+                    lastCommitTime = currentTime;
+                }
+            }
+
             List<SourceRecord> records = task.poll();
-            if (records == null) {
-                return;
+            if (records == null || records.isEmpty()) {
+                traverser = eventTimeMapper.flatMapIdle();
+                emitFromTraverser(traverser);
+                return false;
             }
 
             for (SourceRecord record : records) {
-                boolean added = addToBuffer(record, buf);
-                if (added) {
-                    Map<String, ?> partition = record.sourcePartition();
-                    Map<String, ?> offset = record.sourceOffset();
-                    state.setOffset(partition, offset);
-                }
+                Map<String, ?> partition = record.sourcePartition();
+                Map<String, ?> offset = record.sourceOffset();
+                state.setOffset(partition, offset);
+                task.commitRecord(record);
+            }
+
+            if (commitPeriod == 0) {
+                task.commit();
             }
+
+            traverser = Traversers.traverseIterable(records)
+                    .flatMap(record -> {
+                        T t = map(record);
+                        return t == null ? Traversers.empty() :
+                                eventTimeMapper.flatMapEvent(t, 0, extractTimestamp(record)
+                                );
+                    });
+            emitFromTraverser(traverser);
         } catch (InterruptedException ie) {
             logger.warning("Interrupted while waiting for data");
             Thread.currentThread().interrupt();
         } catch (RuntimeException re) {
             reconnect(re);
         }
+
+        return false;
+    }
+
+    @Nullable
+    protected abstract T map(SourceRecord record);
+
+    private void reconnect(RuntimeException re) {
+        if (reconnectTracker.shouldTryAgain()) {
+            logger.warning("Connection to database lost, will attempt to reconnect and retry operations from " +
+                    "scratch" + getCause(re));
```

Comment:
It might be useful to log the stack trace here and elsewhere where `getCause` is used:
```suggestion
                    "scratch" + getCause(re), re);
```
Currently we don't even have the exception class, just the message and the message of the first cause.

I personally would drop the whole `getCause` method and just rely on the stack trace:
```suggestion
                    "scratch: " + re, re);
```